src/dg/dg.cpp

```diff
@@ -855,7 +855,8 @@ void DGBase<dim,real,MeshType>::assemble_cell_residual (
         }
 
         // CASE 1: FACE AT BOUNDARY
-        if (current_face->at_boundary() && !current_cell->has_periodic_neighbor(iface) ) {
+        if ((current_face->at_boundary() && !current_cell->has_periodic_neighbor(iface)) ||
+            (current_face->at_boundary() && current_cell->has_periodic_neighbor(iface) && dim == 1) ) {//1D periodic BC hardcoded here and confirmed, whereas dealii has shown issues in 1D periodic bc.
```

Comment:
Not since July 2019 and we currently use a version from August 2020.
https://github.com/dealii/dealii/pull/8423

Please use the existing block of code for periodic BCs instead of adding a special 1D case. 
Code block after `//CASE 2: PERIODIC BOUNDARY CONDITIONS`

Reply:
I have found that when using the code block after case 2, for 1D, it does not give the correct results for the energy tests. An error gets introduced. @cpethrick can confirm this with 1D burgers periodic energy tests.
If you'd like, I can add a test showing this phenomenon exactly?

Reply:
This test is passing on my end and I did not need to modify anything in the `/src/` directory

Reply:
@jbrillon Yes that is because your version uses the above snippet of code.
line 865 in
https://github.com/jbrillon/PHiLiP/blob/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed/src/dg/dg.cpp

Reply:
That "||" statement that @dougshidong is referring to, was just added because I noticed that when constructing straight periodic cube, it matches periodic faces for dim=1. Unfortunately, through the energy tests (recovering the exact provable conserved energy), I have found that the dealii match periodic faces introduces errors in 1D, thus this block was added at the start of the PR to solve the issue.
I put that statement that way others don't accidentally use a functionality that we know doesn't behave as expected.

Reply:
What kind of error does it introduce? It can't just affect energy values (unless the test itself isn't set up appropriately), without breaking other periodic solutions. We had other periodic 1D cases in master that never needed this.

Reply:
This is fantastic news! For months/years what I found was that once the test had a couple thousand iterations (that's why I would see it in the energy, and OOA tests on fine grids) then there was a leak of some sort and the solution would start to diverge., but with the periodic BC 1D hard coded it never diverged. I know that Carolyn has also been seeing similar results manifest in her energy tests.
The good news is, I now have it working with the 1D face condition you highlighted above. I am currently finishing up with verifying it passes all 1D tests since deleting that block and will push it soon.
This is great news!

Reply:
Can you clarify what you mean by leak?

So it does work now without adding that extraneous code?

Reply:
1. By leak, what I mean is, before, last tested in June, after a few thousand iterations, the solution would start to diverge. After hard coding the BC, it didn't. For the same test, where the **only** difference was that BC function, if it was more than a few thousand iterations, one would pass while the other diverged. That is why there was the "extraneous code" solely for 1D periodic.
2. Yes, confirming it works now without adding that "extraneous code." Why it magically works now is beyond me. I know that I am not the only one who had this issue, but it seems to be resolved in the current version of this PR. Therefore, it was removed.

Reply:
I did some testing about this issue this morning, and thought I should share. In my draft PR #152, I had built my tests around the fix in this snippet of code (see the note at the end of the PR description). I just reimplemented the hard-coded fix on a local branch, and it demonstrates the energy leak that @AlexanderCicchino is referring to. When I run using the hard-coded 1D periodic BCs, the default deal.ii hypercube grid without periodic BCs, and use a very small timestep for Burgers', energy is conserved to machine precision:

![hard_coded_periodic_boundaries](https://user-images.githubusercontent.com/94074298/179240781-d48602d2-2eeb-4621-b99b-8d58f1163e86.png)

While if I run the same test with `use_periodic_bc=false` and grid from `straight_periodic_cube.cpp`, energy accumulates in an unexpected manner:

![dealii_periodic_boundaries](https://user-images.githubusercontent.com/94074298/179240782-fcf97f2c-d1a0-444a-a7ed-498b667e98f0.png)

It's definitely possible that another change in this PR has somehow fixed the periodic boundary issue, and that my test will run as expected once this PR is added. I'm sharing these results here just to show why this fix was initially implemented, and I think we should still be wary of 1D periodic boundary conditions for energy tests. If new testing by @AlexanderCicchino shows that this is no longer an issue, then that's great! Do we know what could have fixed this bug? 

(unresolving this conversation because I'm interested to know what might have impacted the periodic BCs - we should remove the hard-coded fix if it's no longer needed.)

Reply:
If someone can explain what that fix does differently than without the fix, that would be great. From my understanding, it basically manually fetches the neighbour on the other end of the 1D grid. Is that it?

What happens when you use the periodic neighbor provided by deal.II using Case 2? Note, I'm not asking about the energy, rather what is it fetching differently. Is it somehow the wrong neighbor? The wrong normals? The wrong face? Ignoring the periodic BC? Is the testing failing to assign periodic faces? If so, what kind of boundary condition is it applying?

It's hard for us to understand why that fix is needed based on energy simply increasing. The failure of keeping machine precision energy is a manifestation of another issue.

Reply:
Another factor that might come into play with @cpethrick 's energy numbers is that before the force push the TGV case had some X energy after some time. After the force push, the TGV has some Y energy within 1e-8 relative error from X. Somehow, the strong form implementation loses 8 orders of accuracy for an equivalent implementation.

Worth investigating as a root cause if periodic BC's are correctly assigned, but somehow order of operations affect the your accuracy levels.

Reply:
Thank you very much @cpethrick for showing these results and highlighting the issue perfectly.
@dougshidong Yes, that is exactly it. The fix just manually found the periodic neighbour. 
In both cases the test does identify that it is a periodic BC, and they are fetching the exact same thing! That is why it was extremely weird. They both were passing the exact same cell, solution, dofs, etc to the same assemble face term.
The only difference was that one used dealii periodic face matching to get the face whereas the other was the hardcoded fix. 
I have absolutely no idea why it made a difference, but it did.

I'd argue that failing to keep machine precision energy would not have been another issue because, both tests ran the exact same code, where one used 1 function for just the periodic neighbour, whereas the other used another. If one works perfectly while the other fails, that's a good indicator that the other one was not correct. Could it mean there was an issue with the hardcoded one too? Sure...but the only guarantee I, and all of us, can make with the code is that it passes the tests it has been tested on. Exactly machine precision energy conservation for uncollocated unsteady Burgers with a shock is not something that accidentally happens.
This is also why there was an issue on the master for periodic BCs, and @jbrillon resolved it last month due to this PR.

To conclude: with the current version of this PR that I last pushed, somehow, it fixed whatever issue was at play, and we do not need the snippet of code. I will try to find why because it'd be good to know what change from your master branch allows it to work now, but I have a feeling it's one of those things that are beyond the scope.

Reply:
> Another factor that might come into play with @cpethrick 's energy numbers is that before the force push the TGV case had some X energy after some time. After the force push, the TGV has some Y energy within 1e-8 relative error from X. Somehow, the strong form implementation loses 8 orders of accuracy for an equivalent implementation.
> 
> Worth investigating as a root cause if periodic BC's are correctly assigned, but somehow order of operations affect the your accuracy levels.

The main difference between Burgers and the TGV is that the Burgers enegry result is theoretical and provable, just as the TGV has a theoretical bound on energy provided by the split form test. 
The test you are referring to was not a theoretical energy. Instead it was obtained by running the code, looking at the final result, then inputting that as the answer.
It would be the same as running an OOA test, but instead of looking at the order of convergence, it compares the actual error level value.
For TGV, a 1e-8 difference in energy between code implementations is good enough for the high CFL and coarse grid level the test runs on.

Reply:
Thanks for expanding more on this issue. It would be nice to know why we are experiencing intermittent issues with the periodic BCs, but results accurate to machine precision are good enough to trust that the deal.ii implementation now works with this PR's entropy-stable flux. Resolving the issue (specifically for Burgers' energy) for now is good enough. It would be satisfying to know what the difference is that makes the deal.ii implementation work so that we can avoid issues in the future, but I agree that it could get out of scope quickly! 

In either case, I'm glad that everything is working as expected now.

Reply:
@AlexanderCicchino which PR are you referring to / lines of code? And what was the periodic BC issue I resolved?

Reply:
Could've sworn we had a 1D periodic BC under issues but can't find it so guess I'm wrong about that. Thought it was cleaned up when you went through the other month and resolved a bunch of the issues.

Reply:
Also I don't agree that there should be a difference of 1e-8 in the energy value for the viscous TGV case between the versions of the code before and after your force push for the exact same simulation time, given that the implementation is supposed to be equivalent. In addition, this issue prevails even when the case is ran with twice the number of elements in each direction for the two versions of the code; the difference remained around the same value of ~3.4e-8.

Reply:
Ok I will look into it. As a side note, where can I find the restart file for the restart test?

Reply:
https://github.com/AlexanderCicchino/PHiLiP/pull/23

Reply:
My apologies, I should have been more clear. Would you be able to direct me to the folder where the file is? I was not able to find it in either https://github.com/jbrillon/PHiLiP/tree/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed nor https://github.com/jbrillon/PHiLiP/tree/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed/tests/integration_tests_control_files/taylor_green_vortex_integration nor https://github.com/jbrillon/PHiLiP/tree/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed/src/flow_solver/flow_solver_cases

Reply:
All the parameter files are here:
https://github.com/jbrillon/PHiLiP/tree/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed/tests/integration_tests_control_files/taylor_green_vortex_integration

You have to run `MPI_VISCOUS_TAYLOR_GREEN_VORTEX_RESTART_CHECK` before running `MPI_VISCOUS_TAYLOR_GREEN_VORTEX_RESTART_FROM_PARAMETER_FILE_CHECK` for it to work, that first one dumps the actual solution files that the second test reads in

Reply:
Thank you!

Reply:
@AlexanderCicchino , can you confirm that the current passing test does have a few hundred/thousand iterations to replicate the issue you had seen in the past and @cpethrick 's experience?

> I'd argue that failing to keep machine precision energy would not have been another issue because, both tests ran the exact same code, where one used 1 function for just the periodic neighbour, whereas the other used another. If one works perfectly while the other fails, that's a good indicator that the other one was not correct. Could it mean there was an issue with the hardcoded one too? Sure...but the only guarantee I, and all of us, can make with the code is that it passes the tests it has been tested on. 

The difference is that deal.II actually has unit tests that ensures that the correct cell and face are being fetched whereas we do not. We rely on an integration test that **uses** periodic BC but **checks** energy conservation. We can't infer periodic BC issues from a test that fails due to another criterion. The only thing we can infer from this test is that energy is not conserved.

For all we know, there is a bug in the test set up, and the hard-coded one fixes that. The new code you add may fix energy conservation, but may break periodic BCs for another setting. Of course, it may be the other way around, but  you can't infer it based on energy.

> Exactly machine precision energy conservation for uncollocated unsteady Burgers with a shock is not something that accidentally happens.

I think you underestimate how multiple wrongs can make a right, especially in CFD codes.

> For TGV, a 1e-8 difference in energy between code implementations is good enough for the high CFL and coarse grid level the test runs on.

Also disagree with this. The operators technically only change the order of operations. It is similar to this
```
// Old
x = (a + b + c) * d
```
```
// New
x = c*d + a*d + b*d
```
If the old and new `x` differ by 1e-8, I would be worried. Maybe for some ill-conditioned operations or basis (like monomial), I would lose 2-4 digits. You're doing high-order where the benefits tend to show up in the high-accuracy regions as you get finer and finer. If the same mathematical spatial discretization differs by 1e-8 digits due to two different implementations, one of the implementation is likely erroneous. Maybe the old one was incorrect and the new one is good? Or maybe the other way around?

The overarching theme here is that when you have two processes that **should** give the same results, you can't simply pick the one that gives you the result that you would like to see. You need to understand why one of the process does not give you the desired result. Here's an example from my past experience. I used to verify my automatic differentiation with a finite-difference scheme to build the Jacobian. The finite-difference scheme would give me a Jacobian that was nice and easy for the linear solver and integration tests, whereas the automatic-differentiation Jacobian was singular and could not be used. Turns out there was a BC inconsistency and the automatic-differentiation was the correct Jacobian because it exactly captured that underlying bug. On the other hand, I had a bug in the finite-differences that led to errors to make a non-singular Jacobian. Passing the integration test was more easily done with my erroneous process. While it would be tempting to move on with the finite-difference, it would mean that there are 2 bugs in the code, each of which likely break future students' code. When trying to fix bug A, it would then break the bug B's dependence on bug A, and similarly the other way around. This may manifest itself in 5 years, and by then the expertise or knowledge of the inconsistency within the code would be long gone. I can come up with many different examples similar to this.

Anyways, that's my long-winded opinion on the topic. As long as old tests pass and that we provide test coverage on new code, I wouldn't block a merge. Leaving it up to you and other students to assess risks.

We can resolve this conversation when @cpethrick confirms that the current PR does indeed work through deal.II's periodic faces.

As a sidenote, we should aim to deprecate the use_periodic_bc parameter is we use deal.II's implementation.

Reply:
Confirming the current passing test does have a few thousand (30000) to replicate the issue.
I agree that using dealii's periodic face is more ideal, hence why I changed it to that.
I agree, somewhere there was a bug of some sort, that I have no idea what it was and somehow, somewhere else in this PR it fixed it. 
Yes, again I agree, hence why I've been looking into what is causing the issue with the TGV test all afternoon. 
My only follow-up question is--more for general knowledge--for this TGV test, was the previous answer exactly correct? That is, is there a proof or analytical reason why one value is more correct than the other? I am still working towards replicating the previous result, I'm just curious. My curiosity stems from, if I have an OOA test, let's say 1D linear advection, it I make changes to the code and the error values differ, but the convergent order is still the same, that doesn't necessarily identify which version has the issue, the previous or the recent, or both.
Yes we should deprecate use_periodic_bc parameter.
I feel like we're on the same page about all of this. 
Lastly, I feel like a 5min zoom meeting could clear all of this up for the future!:)

Reply:
> My only follow-up question is--more for general knowledge--for this TGV test, was the previous answer exactly correct? That is, is there a proof or analytical reason why one value is more correct than the other?

Nope. The analytical reason they should be the same is that both implementations reflect the same math.

Yup! I agree that if deal.II's periodic faces gives the same result as your hardcode, there is no bug! :)

Reply:
For this thread, I think we still need @cpethrick  to confirm it works for her other test as well.

Reply:
Yes, I agree that the should be the same for the math side considering the only difference in implementation. Hence why I am happy it passes the test. My question is more, at the current state, is the final kinetic energy what it should be, or will it be changed again in the future? Both of you were quick to establish that there was a 1e-8 difference, but that difference was with the previous kinetic energy value in the prm file. This implies that some changes were made in the code and the 1e-8 difference from the previous commit's parameter file was actually the correct answer.
@cpethrick version is branched from your master branch rather than this PR, so implicitly it shouldn't be working for her because it doesn't have the ~hidden~ fix that's somewhere in this PR.

Reply:
Thank you for the long and well written message @dougshidong, it brings up a very important principle about doing research and I hope lessons will be learned from this discussion.  This is exactly why we cannot allow short cuts / rushed work. 

> My only follow-up question is--more for general knowledge--for this TGV test, was the previous answer exactly correct? That is, is there a proof or analytical reason why one value is more correct than the other?

@AlexanderCicchino Here is the computed kinetic energy for viscous TGV at some time X:
- Weak DG: `1.2073987162646824e-01`
- Strong DG (before the force push): `1.2073987154899971e-01`
This gives a difference of `7.746853e-11`. 

Now, for the exact same test with the new equivalent implementation of operators, there is a difference on the order of `1e-8`, at least 2 orders of magnitude larger than before.

Reply:
> @jbrillon @dougshidong As an update, I have the MPI_VISCOUS_TAYLOR_GREEN_VORTEX_ENERGY_CHECK_LONG test passing locally. The **only** bug was that in the parameter file (tests/integration_tests_control_files/taylor_green_vortex_integration/viscous_taylor_green_vortex_energy_check_long.prm) I was using the one from commit: [18f8ae4](https://github.com/dougshidong/PHiLiP/commit/18f8ae4f7d493d94b6c9c84e5535ef3d7f2c0b15) rather than the latest in commit: [80dc94e#diff-d018221ddce72523c9769b7301c8937eb52e958e0c49540785abc25ea154e75b](https://github.com/dougshidong/PHiLiP/commit/80dc94ee750855b8e952b03f7c8172ed71c13ae6#diff-d018221ddce72523c9769b7301c8937eb52e958e0c49540785abc25ea154e75b) Where the difference was the final time/final kinetic energy/CFL combination. I will push it after merging Doug's latest version of master into it. Fortunately, this means that there is no underlying bug in the code in this PR:)

@AlexanderCicchino @dougshidong  This test is using **weak DG** as seen in the parameter file, this is NOT the failing test we've been discussing the whole time. The failing test is one that I added before you force pushed over my commits. Here is the parameter file https://github.com/jbrillon/PHiLiP/blob/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed/tests/integration_tests_control_files/taylor_green_vortex_integration/viscous_taylor_green_vortex_energy_check_strong_quick.prm

Reply:
@jbrillon , can you confirm that the same end time and CFL have been used to check the resulting energy for strong form before and after. Remember you had this change in the prm file in master

https://github.com/dougshidong/PHiLiP/commit/80dc94ee750855b8e952b03f7c8172ed71c13ae6

Reply:
OK, nvm. The strong form failing is acknowledged to fail. @AlexanderCicchino is working on it.

Reply:
@jbrillon , please post the test name for clarity for the test that changes energy before and after the change.

Reply:
This is the .prm file I have on my branch of this PR for testing weak DG and it passes both before and after the force push:
https://github.com/jbrillon/PHiLiP/blob/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed/tests/integration_tests_control_files/taylor_green_vortex_integration/viscous_taylor_green_vortex_energy_check_weak_quick.prm

Here the .prm file I have for testing strong DG, which passed before the force push and now fails:
https://github.com/jbrillon/PHiLiP/blob/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed/tests/integration_tests_control_files/taylor_green_vortex_integration/viscous_taylor_green_vortex_energy_check_strong_quick.prm

The final time and CFL are the same.

Reply:
`MPI_VISCOUS_TAYLOR_GREEN_VORTEX_ENERGY_CHECK_STRONG_DG_QUICK`

?

Reply:
Yes, that's the name of the test

Reply:
OK. To summarize action items for this convo:

- [ ] @AlexanderCicchino to fix MPI_VISCOUS_TAYLOR_GREEN_VORTEX_ENERGY_CHECK_STRONG_DG_QUICK.
- [ ] @cpethrick to try merging this branch into hers when @AlexanderCicchino fixes conflicts, and to run her tests to confirm periodic BC bug is indeed gone without hard-code.

Reply:
Thank you for listing the test. 
No where in this thread has anyone said the strong form TGV test before that. Only "The TGV case" was said and I thought we were discussing the weak DG long test this entire time because I was comparing to Doug's master branch and it was the one failing. I am working on the strong form TGV case right now.
@dougshidong thank you for the summary.
@jbrillon thank you for your patience.

Reply:
Yeah sorry about not making it more clear that it was strong DG we were talking about, I didn't know you hadn't pulled the latest changes from Doug's master and got it mistaken for that test. These misunderstandings could have been avoided if we had a clean commit history that included my changes -- another example of why we want to avoid force pushing in the future.

Reply:
It now passes with a value of: 0.1207398715510175. Is that ok?

Reply:
The test isn't "passing" if you've changed the expected kinetic energy value of the test to match what you're getting. Before the force push we were getting a value on the order of 1e-11 difference to weak DG, and now we're getting a result that is 1e-7 difference to weak DG with that value you posted above.

Reply:
Sorry but I am confused. Above you posted :

    Weak DG: 1.2073987162646824e-01
    Strong DG (before the force push): 1.2073987154899971e-01
    This gives a difference of 7.746853e-11.

My current result is: 0.1207398715510175
0.1207398715510175 - 1.2073987162646824e-01 = -7.5451e-11
I changed the parameter file's values to match the weak forms and the file in 
https://github.com/jbrillon/PHiLiP/blob/AlexanderCicchino-esfr_split_upstream_update-new-operators-class-force-pushed/tests/integration_tests_control_files/taylor_green_vortex_integration/viscous_taylor_green_vortex_energy_check_strong_quick.prm

Reply:
Sorry about that, I must've made a mistake pasting the values when taking the difference. That's great news, I'm glad you found the bug in strong DG, good work. This is great news; a difference of 2e-12 when compared to the value before the force push.

Reply:
All good :)

Reply:
Oof, and those weren't trivial bugs either.

```cpp
 if(use_auxiliary_eq){
                    local_mass_matrix_aux_state.add(1.0, local_mass_matrix_state);
                }
```

```
            soln_basis_ext.inner_product_surface_1D(neighbor_iface, 
                                                surf_num_flux_minus_surf_soln_ext_dot_normal[istate][idim],
                                                surf_quad_weights, rhs_ext, // <------------- This was rhs_int
                                                soln_basis_ext.oneD_surf_operator,
                                                soln_basis_ext.oneD_vol_operator,
                                                true, 1.0);//it's added since auxiliary is EQUAL to the gradient of the soln
```

Why wouldn't the other tests fail for that? Are we lacking test coverage? A TGV test failing would be hard for anyone other than the implementer to find the root cause.

Reply:
That was somewhat related, but not necessarily. 
The bug that fixed it was in assemble face term auxiliary, there was a rhs_int where it should have been rhs_ext on the surface integral. 
That snippet above that you highlighted was for storing a global inverse mass matrix, but the version was using the inverse on the fly because I hadn't verified the global auxiliary inverse up to that point, but had verified the inverse on-the-fly so it was using that version instead. If that was called before I pushed the last commit, the test would fail immediately with a seg fault. Either way, both inverse on the fly and storing global inverse work for auxiliary now.

Reply:
Sorry yes the second part was the fixed where you highlighted the "This was rhs_int".

Reply:
As a last follow-up, the other tests wouldn't fail for that because the auxiliary variable is dominated by the volume term. The tests for the auxiliary are checking the orders of convergence on a gradient function, and convection diffusion. Since both are those are linear with respect to the auxiliary, the surface integral bug would not have/did not affect the order, for the grid levels tested.

Reply:
Can we have a test that does catch it? Are the convergence levels better for some integration tests? Like being within 0.01 instead of 0.1? Or even better if there is a manufactured solution for which face terms dominate volume terms. Technically, as you refine, I would expect face terms to have more and more impact (maybe?) 

As we clarified, having two energy results of TGV, doesn't tell us much regarding which is right or wrong.

Reply:
The convergence test for conv-diff, I put as 0.1 instead of 0.01 because it requires a finer grid for that finaltime, and would take much too long for its purpose.
Do you have a suggestion on such a manufactured solution? I can't think of one, because from the equation itself, as long as the diffusive coefficient is linear, then the volume term will dominate the orders. The lifting term only comes into play for the orders once the diffusive coefficient becomes nonlinear.
The sole tell-tale test I can think of would be to have a nonlinear diffusive coefficient to render both integrals inexact. Unfortunately, that would require restructuring physics/numerical fluxes to take in b\nabla(u_state) rather than \nabla(u_state). I don't that would be a wise change at this time.

Yes, clarified, thanks!:)

Reply:
Nice. If you're changing code that touches the spatial discretization, it makes sense that you should have tests that covers the spatial discretization.

Definitely worthwhile to make them use strong DG. I guess that right now, we have no tests that verifies that the current strong DG spatial discretization is valid for nonlinear equations?

Reply:
@dougshidong : I found the bug by reading through and checking if everything made sense. The only guarantee that I can make is that the code passes the tests it's been tested on.
@jbrillon @dougshidong after discussing with Doug, we agreed on verifying the Auxiliary equations' terms by checking that the volume and surface integrals in DG Strong recover integration-by-parts exactly. I will work on it now and then push it when it is completed.

Reply:
I have added the unit test, where it verifies DG Strong auxiliary equations RHS for Navier-Stokes for all dim. It does this by setting all modal coefficients to a random number, looping over the cells, calling the dg strong rhs functions and comparing them against their weak form equivalent. 
Also confirming that the test fails if the previous bug gets introduced/any wrong modifications get placed in the rhs volume/surface evaluations.

Reply:
Just updating that I ran the same test by merging onto commit [2f3170e](https://github.com/dougshidong/PHiLiP/pull/111/commits/2f3170ecbc1a7fcfcd71b7becbeaab3941117267) and energy is conserved to machine precision (energy change = 2E-15 using the small timestep as above) using deal.ii periodic boundaries and `conv_num_flux = entropy_conserving_flux`. My result conserving energy indicates that the issue which was causing deal.ii periodic boundaries to fail is no longer impacting 1D Burgers inviscid.